adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,496 @@
+| CUDA interoperability function                                    |  Description
+| [code]#template<backend Backend> +
+platform + 
+make_platform(const backend_input_t<Backend, platform> &backendObject);# 
+        | Create a SYCL `platform` from a list of CUDA device, the list must contain at least one CUDA device.
+
+| [code]#template<backend Backend> +
+device +
+make_device(const backend_input_t<Backend, device> &backendObject);# 
+        | Create a SYCL `device` from a CUDA device.
+
+| [code]#template<backend Backend> +
+context +
+make_context(const backend_input_t<Backend, context> &backendObject,
+                     const async_handler asyncHandler = {});# 
+        | Create a SYCL `context` from a CUDA context.
+
+| [code]#template<backend Backend> +
+queue +
+make_queue(const backend_input_t<Backend, queue> &backendObject,
+                 const context &targetContext,
+                 const async_handler asyncHandler = {});# 
+        | Create a SYCL `queue` from a CUDA stream. The provided `targetContext` must encapsulate the same CUDA context as the provided CUDA stream.
+
+| [code]#template<backend Backend> +
+event +
+make_event(const backend_input_t<Backend, event> &backendObject,
+                 const context &targetContext);# 
+        | Create a SYCL `event` from a CUDA event.
+
+| [code]#template <backend Backend, typename T, int dimensions = 1,
+          typename AllocatorT = buffer_allocator<std::remove_const_t<T>>> +
+buffer<T, dimensions, AllocatorT> +
+make_buffer(const backend_input_t<Backend, buffer<T, dimensions, AllocatorT>>
+               &backendObject,
+           const context &targetContext, event availableEvent);#
+        | Create a SYCL `buffer` from a CUDA device pointer.` The CUDA pointer must be within the provided `targetContext`. The `availableEvent` parameter can be used for synchronization and indicates when the CUDA pointer is ready to be used. Only `dimensions == 1` is supported.
+ 
+| [code]#template <backend Backend, typename T, int dimensions = 1,
+           typename AllocatorT = buffer_allocator<std::remove_const_t<T>>> +
+buffer<T, dimensions, AllocatorT> +
+make_buffer(const backend_input_t<Backend, buffer<T, dimensions, AllocatorT>>
+                &backendObject,
+            const context &targetContext);#
+        | Create a SYCL `buffer` from a CUDA device pointer.` The CUDA pointer must be within the provided `targetContext`. Only `dimensions == 1` is supported.
+
+|====
+
+==== Ownership of native backend objects
+
+The CUDA backend retains ownership of all native CUDA objects obtained through
+the interoperability API, therefore associated SYCL objects must be kept alive
+for the duration of the CUDA work using these native CUDA objects.
+dd CUDA backend specification kernel function interop definitions
+
+When creating a SYCL object from a native CUDA object SYCL does not take
+ownership of the object and it is up to the application to dispose of them when
+appropriate.
```

Comment:
Yes, we looked at that, and I do think it makes sense as well, however it does use an extension header, which is fine as the Level Zero is maintained separately, but it's not ideal if we want the CUDA backend spec to be part of the main SYCL specification.

So we decided to go with a simpler approach for now, for the CUDA backend spec, with the idea in mind that we should probably look into maybe codifying what Level Zero is doing directly in the SYCL interop API, or at least a similar system that could be re-used by the non-OpenCL backends to handle ownership.

Reply:
> however it does use an extension header, which is fine as the Level Zero is maintained separately, but it's not ideal if we want the CUDA backend spec to be part of the main SYCL specification.

Sorry, I'm not following.  What do you mean by "it does use an extension header"?  How does this preclude the interop API from being part of the main SYCL specification?

Reply:
It doesn't preclude it, we just thought it might be better done separately as a generic change to the interop API that could be used by all backends.

Reply:
I'm not sure it makes sense to try to solve this in a general way in the core spec, since only some backends have this issue of "ownership".  (For example, OpenCL doesn't have this issue because objects are reference counted.)  For backends that have this issue (e.g. Level Zero, CUDA), I think it makes sense for each backend to define the mechanism for deciding who "owns" an object.  Of course, it would be better if all backends had a similar mechanism, though.

If the Level Zero mechanism seems good to others, can we use that for CUDA too?  If it has flaws, we could discuss them and potentially adopt a consistent mechanism across both backends.

Reply:
I would argue that the OpenCL backend also has this issue of ownership, it's just that it happens to be easily solved by the OpenCL reference counting. And that the Level Zero approach could easily be implemented by the OpenCL backend, where the ownership transfer would just dictate how SYCL calls `clRetain*` and `clRelease*` on the OpenCL objects.

I could be completely wrong on that but looking at the interop API with `make_*` and `get_*` it seems to me that ownership questions would naturally come up and I get the feeling that maybe it wasn't part of the original interop API simply because it was solved so easily by reference counting for the OpenCL backend.